traffic_portal/app/src/common/modules/table/topologyCacheGroups/TableSelectTopologyCacheGroupsController.js

```diff
@@ -0,0 +1,196 @@
+				usedCacheGroupNames.includes(cg.name); // a cache group that exists in the topology
+		});
+		if (eligibleSecParentCandidates.length === 0) {
+			$uibModalInstance.close({ selectedCacheGroups: selectedCacheGroups, parent: parent.cachegroup, secParent: '' });
+			return;
+		}
+		let params = {
+			title: 'Assign secondary parent?',
+			message: 'Would you like to assign a secondary parent to the following cache groups?<br><br>primary parent = ' + parent.cachegroup + '<br><br>'
+		};
+		params.message += selectedCacheGroups.map(function(cg) { return cg.name }).join('<br>') + '<br><br>';
+		let modalInstance = $uibModal.open({
+			templateUrl: 'common/modules/dialog/confirm/dialog.confirm.tpl.html',
+			controller: 'DialogConfirmController',
+			size: 'md',
+			resolve: {
+				params: function () {
+					return params;
+				}
+			}
+		});
+		modalInstance.result.then(function() {
+			// user wants to select a secondary parent
+			let params = {
+				title: 'Select a secondary parent',
+				message: 'Please select a secondary parent that is part of the ' + topology.name + ' topology',
+				key: 'name'
+			};
+			let modalInstance = $uibModal.open({
+				templateUrl: 'common/modules/dialog/select/dialog.select.tpl.html',
+				controller: 'DialogSelectController',
+				size: 'md',
+				resolve: {
+					params: function () {
+						return params;
+					},
+					collection: function() {
+						// cache groups that are eligible to be a secondary parent include cache groups that are:
+						return eligibleSecParentCandidates;
+					}
+				}
+			});
+			modalInstance.result.then(function(cg) {
+				// user selected a secondary parent
+				$uibModalInstance.close({ selectedCacheGroups: selectedCacheGroups, parent: parent.cachegroup, secParent: cg.name });
+			}, function () {
+				// user apparently changed their mind and doesn't want to select a secondary parent
+				$uibModalInstance.close({ selectedCacheGroups: selectedCacheGroups, parent: parent.cachegroup, secParent: '' });
+			});
+		}, function () {
+			// user doesn't want to select a secondary parent
+			$uibModalInstance.close({ selectedCacheGroups: selectedCacheGroups, parent: parent.cachegroup, secParent: '' });
+		});
+	};
+
+	$scope.cancel = function () {
+		$uibModalInstance.dismiss('cancel');
+	};
+
+	angular.element(document).ready(function () {
+		let availableCacheGroupsTable = $('#availableCacheGroupsTable').DataTable({
+			"scrollY": "60vh",
+			"paging": false,
+			"order": [[ 1, 'asc' ]],
+			"dom": '<"selected-count">frtip',
+			"columnDefs": [
+				{ 'orderable': false, 'targets': [0,5] },
+				{ "width": "5%", "targets": [ 0 ] },
+				{ "width": "35%", "targets": [ 1 ] },
+				{ "width": "15%", "targets": [ 2,3,4,5 ] }
+			],
+			"stateSave": false
+		});
+		availableCacheGroupsTable.on( 'search.dt', function () {
+			$("#selectAllCB").removeAttr("checked"); // uncheck the all box when filtering
```

Comment:
> `s/\$\(([^\)]+)\)/document\.querySelectorAll\(\1\)/g`

This could really just be `document.getElementById('selectAllCB').removeAttribute("checked")`, but that's not really how you're supposed to check and un-check checkboxes, nor is it how you're supposed to manipulate attribute values in Angular. Instead, you should either bind it to a model value in the `$scope` or - slightly worse than that - bind the `checked` attribute to a scope boolean.

Reply:
honestly, i can't remember why i settled on this method (i stole it from some previous code i had written long ago to assign params to profiles). pretty sure i tried a bunch of the angular ways to no avail. probably because i'm trying to work with jquery datatables.

Reply:
The data table event can still be used to toggle the checked state, you'd just update the bound variable instead of directly accessing the DOM. Angular wants to control your DOM access and weird things can happen if you try to do it yourself outside of the digest cycle.

Reply:
> s/\$\(([^\)]+)\)/document\.querySelectorAll\(\1\)/g

sorry, what's this about? i don't read regex :)

Reply:
to paraphrase: down with jQuery. It's unnecessary (except for our datatables) because the functionality that `$` provides is built-in to the browser API

Reply:
i tried
`document.getElementById('selectAllCB').removeAttribute("checked")`
and it didn't work so i think it's fine. yes, there are some jquery uses in TP but not many (outside of datatables)

Reply:
You're really not supposed to load jQuery on the same page as Angularjs at all, because Angularjs ships with its own mini-jQuery and when you mix them you can wind up with events getting swallowed by one and then wonder why you can't get stuff to work in the other. But with the datatables we're locked in until we can get rid of them. I tend to think we should keep jQuery to "only where strictly required" until then.

But also, you actually shouldn't use `document.querySelectorAll` either - Angularjs is built around view/controller separation. Like I said in my original comment:

> _"Instead, you should either bind it to a model value in the `$scope` or - slightly worse than that - bind the `checked` attribute to a scope boolean."_

and I was able to get it work very easily by doing exactly that:

```diff
diff --git a/traffic_portal/app/src/common/modules/table/topologyCacheGroups/TableSelectTopologyCacheGroupsController.js b/traffic_portal/app/src/common/modules/table/topologyCacheGroups/TableSelectTopologyCacheGroupsController.js
index 3ff080965..63c47487a 100644
--- a/traffic_portal/app/src/common/modules/table/topologyCacheGroups/TableSelectTopologyCacheGroupsController.js
+++ b/traffic_portal/app/src/common/modules/table/topologyCacheGroups/TableSelectTopologyCacheGroupsController.js
@@ -22,14 +22,6 @@ var TableSelectTopologyCacheGroupsController = function(parent, topology, cacheG
 	let selectedCacheGroups = [],
 		usedCacheGroupCount = 0;

-	let addAll = function() {
-		markVisibleCacheGroups(true);
-	};
-
-	let removeAll = function() {
-		markVisibleCacheGroups(false);
-	};
-
 	let markVisibleCacheGroups = function(selected) {
 		let visibleCacheGroupIds = $('#availableCacheGroupsTable tr.cg-row').map(
 			function() {
@@ -71,12 +63,10 @@ var TableSelectTopologyCacheGroupsController = function(parent, topology, cacheG
 	});

 	$scope.selectAll = function($event) {
-		const checkbox = $event.target;
-		if (checkbox.checked) {
-			addAll();
-		} else {
-			removeAll();
+		if ($event) {
+			$event.stopPropagation();
 		}
+		markVisibleCacheGroups($scope.allSelected);
 	};

 	$scope.onChange = function(cg) {
@@ -169,6 +159,8 @@ var TableSelectTopologyCacheGroupsController = function(parent, topology, cacheG
 		$uibModalInstance.dismiss('cancel');
 	};

+	$scope.allSelected = false;
+
 	angular.element(document).ready(function () {
 		let availableCacheGroupsTable = $('#availableCacheGroupsTable').DataTable({
 			"scrollY": "60vh",
@@ -184,7 +176,8 @@ var TableSelectTopologyCacheGroupsController = function(parent, topology, cacheG
 			"stateSave": false
 		});
 		availableCacheGroupsTable.on( 'search.dt', function () {
-			$("#selectAllCB").removeAttr("checked"); // uncheck the all box when filtering
+			$scope.allSelected = false;
 		} );
 		decorateCacheGroups();
 		updateSelectedCount();
diff --git a/traffic_portal/app/src/common/modules/table/topologyCacheGroups/table.selectTopologyCacheGroups.tpl.html b/traffic_portal/app/src/common/modules/table/topologyCacheGroups/table.selectTopologyCacheGroups.tpl.html
index 737a98227..0d96b3c67 100644
--- a/traffic_portal/app/src/common/modules/table/topologyCacheGroups/table.selectTopologyCacheGroups.tpl.html
+++ b/traffic_portal/app/src/common/modules/table/topologyCacheGroups/table.selectTopologyCacheGroups.tpl.html
@@ -27,7 +27,7 @@ under the License.
     <table id="availableCacheGroupsTable" class="table responsive-utilities jambo_table" style="table-layout:fixed; width:100%;">
         <thead>
         <tr class="headings">
-            <th style="padding-left: 10px;"><input id="selectAllCB" type="checkbox" ng-click="selectAll($event)"></th>
+            <th style="padding-left: 10px;"><input id="selectAllCB" type="checkbox" ng-change="selectAll($event)" ng-model="allSelected"></th>
             <th>Name</th>
             <th>Type</th>
             <th>Latitude</th>
```

Reply:
ok, i'll try this out. thx.

Reply:
close but doesn't quite work. but close though. specifically this part

`availableCacheGroupsTable.on( 'search.dt', function () {
    $scope.allSelected = false;
 		} );`

Reply:
you did help me identify 2 bugs though :(

i don't know what it is about checkboxes and angularjs but they don't behave as advertised.

Reply:
Really? It worked for me. I'll check it out in Chrome.

Reply:
Oh, I was looking at the wrong thing. This part is for when you _search_ in the table.

This is coincidentally a great example of why loading jQuery in Angular is a bad idea. The event fires, the changes are made, and the `$scope` value is changed - that all happens fine. The problem is, because the event happens in jQuery, Angular is unable to detect the changes and fails to re-digest.

But, it's a data table and so we're forced to hack things to work outside of the Angular digest cycle, then just pray that our changes aren't overwritten if Angular ever needs to redraw anything.

Reply:
turns out i didn't even need this:

`availableCacheGroupsTable.on( 'search.dt', function () { $scope.allSelected = false; } );`

you probably need to let the jquery thing go. there is some jquery in TP. it works and has worked for a long time. of course, if there's a better angularjs way to do things, they should be done but replacing 1 jquery line with 20 lines of vanilla javascript is not a good tradeoff imo especially for something that works.

Reply:
Well I wouldn't recommend 20 lines of vanilla js either; just 3 lines of AngularJS (and sometimes it even saves LOC to do things the Angular way, depending on what it is). The problem is that loading jQuery prevents you from doing Angular things - which isn't good when you're building an Angular UI.

Data tables are here for now, but that doesn't necessarily mean they're here to stay - esp. as we migrate away from AngularJS to proper Angular. But that's a conversation that we don't need to have here.

In this particular case the only way to handle it is with jQuery. It's just regrettable that doing so results in untestable and inconsistent behavior.